From 8feb8b4b8634790ff6ed7e4e54fbe3eeaa14aba8 Mon Sep 17 00:00:00 2001
From: Matteo Bruni <mbruni@codeweavers.com>
Date: Mon, 6 Oct 2025 17:44:14 +0000
Subject: [PATCH 1/2] dsound: Set all channel volumes to 0 when DSBVOLUME_MIN
 is passed in.

Notably from DSOUND_Calc3DBuffer() for distant sounds.

The idea is to make secondarybuffer_is_audible() work with > 2 channels.
---
 dlls/dsound/mixer.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/dlls/dsound/mixer.c b/dlls/dsound/mixer.c
index dbc84cbd04f..f7192bfb1b6 100644
--- a/dlls/dsound/mixer.c
+++ b/dlls/dsound/mixer.c
@@ -50,6 +50,13 @@ void DSOUND_RecalcVolPan(PDSVOLUMEPAN volpan)
 	TRACE("Vol=%ld Pan=%ld\n", volpan->lVolume, volpan->lPan);
 	/* the AmpFactors are expressed in 16.16 fixed point */
 
+	if (volpan->lVolume == DSBVOLUME_MIN)
+	{
+		for (unsigned int i = 0; i < DS_MAX_CHANNELS; i++)
+			volpan->dwTotalAmpFactor[i] = 0;
+		TRACE("setting all channel volumes to 0\n");
+		return;
+	}
 	/* FIXME: use calculated vol and pan ampfactors */
 	temp = (double) (volpan->lVolume - (volpan->lPan > 0 ? volpan->lPan : 0));
 	volpan->dwTotalAmpFactor[0] = (ULONG) (pow(2.0, temp / 600.0) * 0xffff);

From a170f9d6e9dc6b81da319e30fe1ec5cb934d79c7 Mon Sep 17 00:00:00 2001
From: Matteo Bruni <mbruni@codeweavers.com>
Date: Mon, 24 Nov 2025 21:43:44 +0100
Subject: [PATCH 2/2] dsound: Flush denormals in the mixing thread.

I suspect denormals might be the cause of the unexplained slowness
mentioned in a few of the later comments on bug 30639. Denormals can
be much slower than normal floats, although the penalty is generally
significant only on older CPUs. 64-bit processes default to flushing,
so this only really matters for 32-bit.
---
 dlls/dsound/mixer.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/dsound/mixer.c b/dlls/dsound/mixer.c
index f7192bfb1b6..623c15496a8 100644
--- a/dlls/dsound/mixer.c
+++ b/dlls/dsound/mixer.c
@@ -767,8 +767,10 @@ static void DSOUND_PerformMix(DirectSoundDevice *device)
 DWORD CALLBACK DSOUND_mixthread(void *p)
 {
 	DirectSoundDevice *dev = p;
+
 	TRACE("(%p)\n", dev);
 	SetThreadDescription(GetCurrentThread(), L"wine_dsound_mixer");
+        _controlfp_s(NULL, _DN_FLUSH, _MCW_DN);
 
 	while (dev->ref) {
 		DWORD ret;
